qt/lib/include/verovio-qt/veroviodoc.h

```diff
@@ -0,0 +1,208 @@
+
+#include <QObject>
+#include <QVector>
+
+//----------------------------------------------------------------------------
+
+#include "toolkit.h"
+
+namespace vrv {
+
+//----------------------------------------------------------------------------
+// VerovioDoc
+//----------------------------------------------------------------------------
+
+/**
+ * This class represents one Verovio document.
+ */
+class VerovioDoc : public QObject {
+    Q_OBJECT
+
+    Q_PROPERTY(QString resourcesDataPath MEMBER m_resourcesDataPath WRITE setResourcesDataPath)
+    Q_PROPERTY(QString fontDirPath MEMBER m_fontDirPath WRITE setFontDirPath)
+    Q_PROPERTY(QString fileName MEMBER m_fileName WRITE setFileName)
+    Q_PROPERTY(QString fileContent MEMBER m_fileContent WRITE setFileContent)
+    Q_PROPERTY(QString musicFont MEMBER m_musicFont WRITE setMusicFont)
+    Q_PROPERTY(int pageCount MEMBER m_pageCount READ pageCount NOTIFY pageCountChanged)
+    Q_PROPERTY(int pageWidth MEMBER m_pageWidth WRITE setPageWidth)
+    Q_PROPERTY(int pageHeight MEMBER m_pageHeight WRITE setPageHeight)
+    Q_PROPERTY(int scale MEMBER m_scale WRITE setScale)
+    Q_PROPERTY(int border MEMBER m_border WRITE setBorder)
+    Q_PROPERTY(int spacingStaff MEMBER m_spacingStaff WRITE setSpacingStaff)
+    Q_PROPERTY(int spacingSystem MEMBER m_spacingSystem WRITE setSpacingSystem)
+    Q_PROPERTY(bool adjustPageHeight MEMBER m_adjustPageHeight WRITE setAdjustPageHeight)
+    Q_PROPERTY(bool noLayout MEMBER m_noLayout WRITE setNoLayout)
+    Q_PROPERTY(bool ignoreLayout MEMBER m_ignoreLayout WRITE setIgnoreLayout)
+    Q_PROPERTY(bool hasValidData MEMBER m_hasValidData READ hasValidData NOTIFY hasValidDataChanged)
+
+public:
+    /**
+     * @name Constructors, destructors, and other standard methods
+     */
+    ///@{
+    VerovioDoc();
+    virtual ~VerovioDoc() = default;
+    ///@}
+
+    /**
+     * @name Getter/setter for the adjusted page height if the adjustPageHeight is activated.
+     */
+    ///@{
+    Q_INVOKABLE int adjustedPageHeightForPage(int page);
+    void setAdjustedPageHeightForPage(int pageNumber, int pageHeight);
+    ///@}
+
+    /**
+     * @name Getter for the page count.
+     */
+    int pageCount() const { return m_pageCount; }
+
+    /**
+     * @name Render the given page of this document to the specified deviceContext.
+     */
+    void renderPage(int page, DeviceContext *deviceContext);
+
+    /**
+     * @name Returns if the document has valid data to show.
+     */
+    bool hasValidData() const { return m_hasValidData; }
+
+public slots:
+    /**
+     * @name Setters for public properties.
+     */
+    ///@{
+    void setResourcesDataPath(QString resourcesDataPath);
+    void setFontDirPath(QString fontDirPath);
+    void setFileName(QString fileName);
+    void setFileContent(QString fileContent);
+    void setMusicFont(QString musicFont);
+    void setPageWidth(int pageWidth);
```

Comment:
Why not having one single method for a this, namely setOptions that takes a JSON string an pass it to the vrv::toolkit?

Reply:
These setters are directly used by the Q_PROPERTY definition which exposes the fields as QML attributes. If using a JSON string instead, constructing the required options from QML would be a lot more cumbersome.
See: http://doc.qt.io/qt-5/qtqml-cppintegration-exposecppattributes.html